agent/consul/multilimiter/multilimiter.go

```diff
@@ -0,0 +1,128 @@
+package multilimiter
+
+import (
+	"context"
+	radix "github.com/hashicorp/go-immutable-radix"
+	"golang.org/x/time/rate"
+	"sync/atomic"
+	"time"
+)
+
+type LimitedEntity interface {
+	Key() []byte
+}
+
+type RateLimiter interface {
+	Start()
+	Stop()
+	Allow(entity LimitedEntity) bool
+	UpdateConfig(c Config)
+}
+
+type Limiter struct {
+	limiter    *rate.Limiter
+	lastAccess atomic.Int64
+}
+
+type Config struct {
+	Rate         rate.Limit
+	Burst        int
+	CleanupLimit time.Duration
+	CleanupTick  time.Duration
+}
+
+type MultiLimiter struct {
+	limiters *atomic.Pointer[radix.Tree]
+	config   *atomic.Pointer[Config]
+	cancel   context.CancelFunc
+}
+
+func (m *MultiLimiter) UpdateConfig(c Config) {
```

Comment:
I thought you mentioned this was going away?

Reply:
Would that be problematic when configuration has changed but the server/ip/tenant/service is read more often than the CleanupLimit?   It seems like it would never update while this is true.

Reply:
yes I thought about it more and I think the 2 options we have are:
- update when allow is called
- update out of band in the clean up loop
I implemented the first option but I think we need to rework it as this mean that on config update all the new requests will go the slow path which is not ideal. 
Ideally we will do it out of band to avoid having rate limited requests go through  the slow path.
I think this can be done later, as for global limiter it's a no issue (we will only have 1 limiter for all the connections).